File: mod-common/src/main/java/com/dreamdisplays/net/DisplayInfoPacket.java
package com.dreamdisplays.net;

import org.joml.Vector3i;
import com.dreamdisplays.PlatformlessInitializer;
import com.dreamdisplays.util.Facing;

import java.util.UUID;
import net.minecraft.core.UUIDUtil;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;

// Packet for sending display information
public record DisplayInfoPacket(UUID id, UUID ownerId, Vector3i pos, int width, int height, String url, Facing facing, boolean isSync, String lang) implements CustomPacketPayload {
    public static final CustomPacketPayload.Type<DisplayInfoPacket> PACKET_ID =
            new CustomPacketPayload.Type<>(Identifier.fromNamespaceAndPath(PlatformlessInitializer.MOD_ID, "display_info"));

    public static final StreamCodec<FriendlyByteBuf, DisplayInfoPacket> PACKET_CODEC =
            StreamCodec.of(
                    (buf, packet) -> {
                        UUIDUtil.STREAM_CODEC.encode(buf, packet.id());
                        UUIDUtil.STREAM_CODEC.encode(buf, packet.ownerId());

                        ByteBufCodecs.VAR_INT.encode(buf, packet.pos().x());
                        ByteBufCodecs.VAR_INT.encode(buf, packet.pos().y());
                        ByteBufCodecs.VAR_INT.encode(buf, packet.pos().z());

                        ByteBufCodecs.VAR_INT.encode(buf, packet.width());
                        ByteBufCodecs.VAR_INT.encode(buf, packet.height());

                        ByteBufCodecs.STRING_UTF8.encode(buf, packet.url());

                        ByteBufCodecs.BYTE.encode(buf, packet.facing().toPacket());
                        ByteBufCodecs.BOOL.encode(buf, packet.isSync());

                        ByteBufCodecs.STRING_UTF8.encode(buf, packet.lang());
                    },
                    (buf) -> {
                        UUID id = UUIDUtil.STREAM_CODEC.decode(buf);
                        UUID ownerId = UUIDUtil.STREAM_CODEC.decode(buf);

                        int x = ByteBufCodecs.VAR_INT.decode(buf);
                        int y = ByteBufCodecs.VAR_INT.decode(buf);
                        int z = ByteBufCodecs.VAR_INT.decode(buf);
                        Vector3i pos = new Vector3i(x, y, z);

                        int width = ByteBufCodecs.VAR_INT.decode(buf);
                        int height = ByteBufCodecs.VAR_INT.decode(buf);

                        String url = ByteBufCodecs.STRING_UTF8.decode(buf);

                        byte facingByte = ByteBufCodecs.BYTE.decode(buf);
                        Facing facing = Facing.fromPacket(facingByte);

                        boolean isSync = ByteBufCodecs.BOOL.decode(buf);
                        String lang = ByteBufCodecs.STRING_UTF8.decode(buf);

                        return new DisplayInfoPacket(id, ownerId, pos, width, height, url, facing, isSync, lang);
                    }
            );

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return PACKET_ID;
    }
}

File: common/src/main/kotlin/com/dreamdisplays/net/s2c/DisplayInfoPacket.kt
package com.dreamdisplays.net.s2c

import com.dreamdisplays.net.common.helpers.createType
import com.dreamdisplays.util.Facing
import com.dreamdisplays.util.Facing.Companion.fromPacket
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.codec.StreamCodec.of
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.network.protocol.common.custom.CustomPacketPayload.Type
import org.joml.Vector3i
import java.util.*

data class DisplayInfoPacket(
    val uuid: UUID,
    val ownerUuid: UUID,
    val pos: Vector3i,
    val width: Int,
    val height: Int,
    val url: String,
    val facing: Facing,
    val isSync: Boolean,
    val lang: String,
) : CustomPacketPayload {
    override fun type(): Type<out CustomPacketPayload> = PACKET_ID

    companion object {
        val PACKET_ID: Type<DisplayInfoPacket> = createType("display_info")
        val PACKET_CODEC: StreamCodec<FriendlyByteBuf, DisplayInfoPacket> = of(
            { buf, packet ->
                buf.writeUUID(packet.uuid)
                buf.writeUUID(packet.ownerUuid)
                buf.writeVarInt(packet.pos.x())
                buf.writeVarInt(packet.pos.y())
                buf.writeVarInt(packet.pos.z())
                buf.writeVarInt(packet.width)
                buf.writeVarInt(packet.height)
                buf.writeUtf(packet.url)
                buf.writeByte(packet.facing.toPacket().toInt())
                buf.writeBoolean(packet.isSync)
                buf.writeUtf(packet.lang)
            },
            { buf ->
                DisplayInfoPacket(
                    buf.readUUID(),
                    buf.readUUID(),
                    Vector3i(
                        buf.readVarInt(),
                        buf.readVarInt(),
                        buf.readVarInt()
                    ),
                    buf.readVarInt(),
                    buf.readVarInt(),
                    buf.readUtf(),
                    buf.readByte().fromPacket(),
                    buf.readBoolean(),
                    buf.readUtf()
                )
            }
        )
    }
}
